File: c#/42-Trapping-Rain-Water.cs
// Water is going to be bound by lower side
// If left height is less than right, water is bound by left, so calculate any trapped water & add to sum by subtracting curr left height from max left height, then iterate left ptr forward
// If right side is lower, do the above for right instead
// Once left > right, return sum

// TWO POINTER SOLUTION
// O(N) Time, O(1) Space

public class Solution {
    public int Trap(int[] height) {
        int left = 0, right = height.Length - 1;

        int leftMax = height[left], rightMax = height[right];

        int sum = 0;

        while (left <= right)
        {
            int leftCur = height[left];
            int rightCur = height[right];
            // Water is bound by left, calc water trapped by leftMax
            if (leftCur < rightCur)
            {
                leftMax = Math.Max(leftMax, leftCur);
                sum += leftMax - leftCur;
                left++;
            }
            else
            {
                rightMax = Math.Max(rightMax, rightCur);
                sum += rightMax - rightCur;
                right--;
            }
        }

        return sum;
    }
}

File: python/42-Trapping-Rain-Water.py
from typing import List

# region Dynamic Programming
class Solution:
    def trap(self, height: List[int]) -> int:
        """
        Dynamic Programming

        * Get max elevation seen from the left and right for each index
        * At each height except for the ends, subtract the cur height from the min of the left and right max elev values seen for that idx
        * If the diff is greater than 0 add it to the retVal

        Time O(n)
        Space O(n)
        """
        leftElev, rightElev = [0] * len(height), [0] * len(height)

        for i in range(1, len(height)):
            leftElev[i] = max(height[i - 1], leftElev[i - 1])
            j = len(height) - i - 1
            rightElev[j] = max(height[j + 1], rightElev[j + 1])

        retVal = 0
        for i in range(1, len(height) - 1):
            water = max(0, min(leftElev[i], rightElev[i]) - height[i])
            retVal += water
        return retVal


# endregion

# region Two Pointers
class Solution:
    def trap(self, height: List[int]) -> int:
        """
        Two Pointers

        Start two pointers at either end of the array
        Util the pointers haven't passed each other, do the following:
          * Water is bound by the smaller of the two current values (leftCur and rightCur)
          * For the one which is smaller, compare it to the max seen for that side
          * If it's smaller than the max, add diff to the retVal and iterate the pointer toward the center

        Time O(n)
        Space O(1)

        """
        left, right = 0, len(height) - 1

        leftMax, rightMax = height[left], height[right]

        retVal = 0

        while left <= right:
            leftCur = height[left]
            rightCur = height[right]

            if leftCur < rightCur:
                leftMax = max(leftMax, leftCur)
                retVal += leftMax - leftCur
                left += 1
            else:
                rightMax = max(rightMax, rightCur)
                retVal += rightMax - rightCur
                right -= 1

        return retVal


# endregion
